experiment/cherrypicker/server.go

```diff
@@ -0,0 +1,197 @@
+	if !ic.Issue.IsPullRequest() || ic.Issue.State != "closed" || ic.Action != github.IssueCommentActionCreated {
+		return nil
+	}
+
+	cherryPickMatches := cherryPickRe.FindAllStringSubmatch(ic.Comment.Body, -1)
+	if len(cherryPickMatches) == 0 || len(cherryPickMatches[0]) != 2 {
+		return nil
+	}
+	targetBranch := cherryPickMatches[0][1]
+
+	org := ic.Repo.Owner.Login
+	repo := ic.Repo.Name
+	num := ic.Issue.Number
+	commentAuthor := ic.Comment.User.Login
+
+	// TODO: Filter out notifications from repos we don't want to do cherry-picks.
+
+	pr, err := s.GithubClient.GetPullRequest(org, repo, num)
+	if err != nil {
+		return err
+	}
+	baseBranch := pr.Base.Ref
+
+	// TODO: Use a whitelist for allowed base and target branches.
+	if baseBranch != "master" || baseBranch == targetBranch {
+		resp := fmt.Sprintf("Base branch (%s) needs to be 'master' and differ from target branch (%s)", baseBranch, targetBranch)
+		s.Logger.Info(resp)
+		return s.GithubClient.CreateComment(org, repo, num, plugins.FormatResponseRaw(ic.Issue.Body, ic.Issue.HTMLURL, commentAuthor, resp))
+	}
+
+	// Only members should be able to do cherry-picks.
+	ok, err := s.GithubClient.IsMember(org, commentAuthor)
+	if err != nil {
+		return err
+	}
+	if !ok {
+		resp := fmt.Sprintf("only %s org members may request cherry picks", org)
+		s.Logger.Info(resp)
+		return s.GithubClient.CreateComment(org, repo, num, plugins.FormatResponseRaw(ic.Issue.Body, ic.Issue.HTMLURL, commentAuthor, resp))
+	}
+
+	// Cherry-pick only merged PRs.
+	if !pr.Merged {
+		resp := "cannot cherry-pick an unmerged PR"
+		s.Logger.Info(resp)
+		return s.GithubClient.CreateComment(org, repo, num, plugins.FormatResponseRaw(ic.Issue.Body, ic.Issue.HTMLURL, commentAuthor, resp))
+	}
+
+	// Clone the repo, checkout the target branch.
+	startClone := time.Now()
+	r, err := s.GitClient.Clone(ic.Repo.FullName)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := r.Clean(); err != nil {
+			s.Logger.WithError(err).Error("Error cleaning up repo.")
+		}
+	}()
+	if err := r.Checkout(targetBranch); err != nil {
+		return err
+	}
+	s.Logger.WithField("duration", time.Since(startClone)).Info("Cloned and checked out target branch.")
+
+	// Fetch the patch from Github
+	rsp, err := http.Get(fmt.Sprintf("https://patch-diff.githubusercontent.com/raw/%s/%s/pull/%d.patch", org, repo, num))
+	if err != nil {
+		return err
+	}
+	defer rsp.Body.Close()
+	if rsp.StatusCode < 200 || rsp.StatusCode > 299 {
+		return fmt.Errorf("response status was not 2XX: %s", rsp.Status)
+	}
+	localPath := fmt.Sprintf("/tmp/%d-%s.patch", num, uuid.NewV1().String())
+	out, err := os.Create(localPath)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+	if _, err := io.Copy(out, rsp.Body); err != nil {
+		return err
+	}
+
+	if err := r.Config("user.name", "cherrypicker"); err != nil {
+		return err
+	}
+	if err := r.Config("user.email", "cherrypicker@localhost"); err != nil {
+		return err
+	}
+
```

Comment:
check if the patch has already been cherry-picked to target branch?

Reply:
Trying to apply the same patch more than once fails with
```
Applying: Another test
error: patch failed: new_file:1
error: new_file: patch does not apply
Patch failed at 0001 Another test
The copy of the patch that failed is found in: .git/rebase-apply/patch
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort".
```

Reply:
what if someone can't see the log? Assume this case, someone else cherry-picked the PR manually and didn't xref so I don't know this. I try the command but nothing will happen. I can't know why it doesn't work. I'd like the bot give me a comment on this.

Reply:
You would get the comment if I hadn't push my own cherry-pick already. In your case, it failed in a previous step (the cherry-pick branch already existed on the fork).

Reply:
Sorry, I am conflating things. The gist is that you will get a comment if apply fails.

Reply:
Won't that get surfaced in L173?

Reply:
Yes